### ISSUE TITLE
Similaries with Leela Zero

### ISSUE THREAD

**Issue by as0770** [Chicoryn/dream-go#27]
Just observed some similaries with Leela Zero. Is Dream Go some kind of rewritten Leela? What is taken from Leela and where are the changes?

**Comment by Chicoryn** [comment]
Nothing has been taken from Leela, however there is certainly some inspiration coming from a lot of different go related project. Some examples of inspirations from different projects:

* The console output is heavily inspired by leela in order to be able to re-use some wrappers like [SabakiLeela](https://github.com/SabakiHQ/LeelaSabaki), and because there was no need to re-invent the wheel.
* Some concepts like [FPU reduction](https://github.com/gcp/leela-zero/issues/696) are implemented in Dream Go too. There are some other things that has been discovered in parallel such as the fact that one can merge batch normalization directly into the weights #9 (that the Leela team only discovered recently).
* Some of the game state code were inspired by how [GNU Go](https://github.com/liangsirrun/gnugo) represents their game state.
* The docker images and game database code were inspired by [minigo](https://github.com/tensorflow/minigo).

Beyond inspiration (going both ways I suspect, since a lot of changes we have made has been suggested a week or so later in their issues) Leela Zero and Dream Go are completely independent projects and, as far as I know, share no code. But both projects are based on the same research paper [[1]](https://deepmind.com/documents/119/agz_unformatted_nature.pdf) published by Google DeepMind so there are plenty of similarities because of that, for example, both use the same MCTS algorithm to find the best move and roughly the same neural network architecture (ResNet).

There are still some differences, the main ones would be:

* Dream Go tries to cheat as much as possible to gain strength, by using things like ladder specific features, while Leela Zero is a more faithful adaptation of what DeepMind implemented.

   Since neither project is very far into their learning process so far we don't yet know whether it is worth it to _"cheat"_, or if it detrimental in the long run, but maybe we will know in a year or two.
* Dream Go makes use of _Expert Iteration_ [[2]](https://arxiv.org/pdf/1705.08439.pdf) for reinforcement learning instead of the _AlphaZero_ [[1]](https://deepmind.com/documents/119/agz_unformatted_nature.pdf) reinforcement learning algorithm that Leela Zero use.

   This is because of necessity, Leela Zero has about 488 times my compute resources (at the time of writing this post) and if you check their strength graph you can see I would be nowhere if I had used the same approach (this applies mostly to 0.6.0, the 0.5.0 weights are only based on human games). Most likely _Expert Iteration_ is an inferior algorithm to _AlphaZero_ but there is no real choice here.
* Both Dream Go and Leela Zero generate their own independent training data, so their neural networks may have learned completely different things. With different strengths and weaknesses, or maybe one just being flat-out stronger than the other (The latest _Leela Zero_ is probably stronger than _Dream Go v0.5.0_, I have not tested the latest release candidate for _v0.6.0_).

The rest of the differences are implementation details, such as _Dream Go_ being written in Rust and _Leela Zero_ in C++.

[1] https://deepmind.com/documents/119/agz_unformatted_nature.pdf
[2] https://arxiv.org/pdf/1705.08439.pdf

**Comment by as0770** [comment]
Many thanks for the detailed answer! It is very interesting to know some details like this. 

In fact I was refering  to the console output. I wasn't aware that this might be some feature for a special GUI. I played a match with twogtp in console and saw this:

`LeelaZero:
Playouts: 3383, Win: 90.58%, PV: L5 C15 G17 J17 C16 D15 C11 E17 E18 E16 C18 G16 F17 F16 H17 H16 J18
  DreamGo:
Nodes: 1611, Win: 38.3%, PV: n7 n8 m9 m8 c15 g17 j17 c16 d15 c12 g15 h16 j14 j16

LeelaZero:
 L5 ->    4438 (V: 91.57%) (N: 18.83%) PV: L5 C15 G17 J17 C16 D15 C11 E17 E18 E16 C18 G16 F17 F16 H17 H16 J18
  L2 ->     493 (V: 90.92%) (N:  7.95%) PV: L2 N8 N9 M9 M8 M10 N7 C15 G17 J17 C16 D15 C11 E17
  N7 ->     142 (V: 79.50%) (N: 30.10%) PV: N7 L5 K5 K4 O6 P6 K3 K6 L7 K4 P5 O6 K3 O8 N9 K4
  N8 ->      50 (V: 80.71%) (N:  9.38%) PV: N8 L5 L2 K6 J7 K5 K4 K7
  M9 ->      33 (V: 84.37%) (N:  3.96%) PV: M9 L5 K4 K6 J7 L7 K7 N8 N9 M8
  O8 ->      26 (V: 79.63%) (N:  5.45%) PV: O8 L5 K5 K4 N7 O6 K3 K6 L7 K4
 J17 ->      19 (V: 78.66%) (N:  4.26%) PV: J17 L5 L2 K6 L7 K5 K4 J7
  O3 ->      12 (V: 81.99%) (N:  2.15%) PV: O3 L5 K5 K4 O4 P5
  N9 ->      10 (V: 80.63%) (N:  1.83%) PV: N9 L5 K5 K4 L2 K3
 D14 ->       9 (V: 80.87%) (N:  1.79%) PV: D14 L5 K5 K4 J17
 H17 ->       7 (V: 79.82%) (N:  1.27%) PV: H17 L5 K5 K4
 D15 ->       7 (V: 77.31%) (N:  1.73%) PV: D15 L5 K5 K4 N7
  O6 ->       6 (V: 83.08%) (N:  1.01%) PV: O6 L5 K5 K4
 K17 ->       6 (V: 79.79%) (N:  1.29%) PV: K17 L5 K5 K4 K15
  O4 ->       6 (V: 75.36%) (N:  1.93%) PV: O4 O3 P3 L5 K5 K4
 M14 ->       5 (V: 83.02%) (N:  0.76%) PV: M14 L5 K5 K4
 N10 ->       4 (V: 83.37%) (N:  0.56%) PV: N10 L5 K5 K4
 C14 ->       4 (V: 81.44%) (N:  0.56%) PV: C14 L5 K5 K4
  P5 ->       4 (V: 76.42%) (N:  0.99%) PV: P5 L5 K5 K4
5282 visits, 1338942 nodes, 3615 playouts, 225 n/s

DreamGo:
   n7 ->    1119 (W: 39.45%) (N: 43.10%) PV: n7 n8 m9 m8 c15 g17 j17 c16 d15 c12 g15 h16 j14 j16
   n8 ->     306 (W: 37.22%) (N: 16.75%) PV: n8 n9 m9 m8 n7 m10 l9 n11 k8 l7 k9 l2 n13 k11 h10
   n9 ->      71 (W: 36.68%) (N:  5.02%) PV: n9 n8 o8 o10 n7 m8 p5 o11
   p5 ->      48 (W: 33.06%) (N:  8.04%) PV: p5 n7 o6 o8 d15 g17 c17 c18 c16 k17 b18
   o8 ->      14 (W: 30.25%) (N:  2.92%) PV: o8 n9 p5 p6 o6 o3 o4 l2
  m12 ->      12 (W: 33.75%) (N:  2.04%) PV: m12 k17 n8 n9 m9 m10
   s1 ->      11 (W: 32.94%) (N:  1.96%) PV: s1 q1 n8 n9 m8
   o3 ->       6 (W: 22.85%) (N:  2.82%) PV: o3 n7 o6 o8 p5 p3 o4
  e12 ->       6 (W: 27.15%) (N:  0.96%) PV: e12 g17
   m9 ->       5 (W: 28.97%) (N:  1.36%) PV: m9 n8 n7 m8 p5 p6
   k8 ->       2 (W: 24.66%) (N:  1.12%) PV: k8 l9 d15
`
